src/frontend/Deprecation_analysis.ml

```diff
@@ -0,0 +1,168 @@
+    (Location_span.to_string loc_span)
+    message
+
+let replace_suffix = function
+  | { stmt= FunDef {funname= {name; _}; arguments= (_, type_, _) :: _; _}
+    ; smeta= _ }
+    when String.is_suffix ~suffix:"_log" name ->
+      let newname =
+        if String.is_suffix ~suffix:"_cdf_log" name then
+          String.drop_suffix name 8 ^ "_lcdf"
+        else if String.is_suffix ~suffix:"_ccdf_log" name then
+          String.drop_suffix name 9 ^ "_lccdf"
+        else if Middle.UnsizedType.is_real_type type_ then
+          String.drop_suffix name 4 ^ "_lpdf"
+        else String.drop_suffix name 4 ^ "_lpmf"
+      in
+      String.Table.add deprecated_userdefined ~key:name ~data:newname
+      |> (ignore : [`Ok | `Duplicate] -> unit)
+  | _ -> ()
+
+let rec warn_deprecated_expr
+    ({expr; emeta} : (typed_expr_meta, fun_kind) expr_with) :
+    (typed_expr_meta, fun_kind) expr_with =
+  let expr =
+    match expr with
+    | GetLP ->
+        warn_deprecated
+          ( emeta.loc
+          , "The no-argument function `get_lp()` is deprecated. Use the \
+             no-argument function `target()` instead." ) ;
+        GetLP
+    | FunApp (StanLib, {name= "abs"; id_loc}, [e])
+      when Middle.UnsizedType.is_real_type e.emeta.type_ ->
+        warn_deprecated
+          ( emeta.loc
+          , "Use of the `abs` function with real-valued arguments is \
+             deprecated; use functions `fabs` instead." ) ;
+        FunApp (StanLib, {name= "abs"; id_loc}, [warn_deprecated_expr e])
+    | FunApp (StanLib, {name= "if_else"; id_loc}, l) ->
+        warn_deprecated
+          ( emeta.loc
+          , "The function `if_else` is deprecated. Use the conditional \
+             operator (x ? y : z) instead." ) ;
+        FunApp
+          ( StanLib
+          , {name= "if_else"; id_loc}
+          , List.map ~f:warn_deprecated_expr l )
+    | FunApp (StanLib, {name; id_loc}, l) ->
+        if Option.is_some (String.Map.find deprecated_distributions name) then
+          warn_deprecated
+            ( emeta.loc
+            , name ^ " is deprecated, use " ^ rename_distribution name
+              ^ " instead" )
+        else if Option.is_some (String.Map.find deprecated_functions name) then
+          warn_deprecated
+            ( emeta.loc
+            , name ^ " is deprecated, use " ^ rename_function name ^ " instead"
+            ) ;
+        FunApp (StanLib, {name; id_loc}, List.map ~f:warn_deprecated_expr l)
+    | FunApp (UserDefined, {name; id_loc}, l) ->
+        let newname = String.Table.find deprecated_userdefined name in
+        if Option.is_some newname then
+          warn_deprecated
+            ( emeta.loc
+            , "Use of the _log suffix in user defined function " ^ name
+              ^ " is deprecated, use " ^ Option.value_exn newname ^ " instead."
+            ) ;
+        FunApp (UserDefined, {name; id_loc}, List.map ~f:warn_deprecated_expr l)
+    | _ -> map_expression warn_deprecated_expr ident expr
```

Comment:
Hmm, there may be another way to iterate over the tree for warnings without needing to return the tree. Maybe fold_expression et al with unit accumulator?

Reply:
I like the idea but I'm not sure how to implement it. At the moment there's no `fold_expression`, right (the closest I've found is `fold_expr` in `Mir_utils`)? Does this mean changing expression's `@@deriving` statement somehow?

Reply:
Hmm, according to the doc: https://github.com/ocaml-ppx/ppx_deriving it looks like there's even an `iter` method. Though I will say it's better functional style to accumulate the error messages rather than printing them out procedurally when they are encountered. For the procedural style you could add `iter` to all of the relevant `@@deriving` statements in Ast.ml; for functional style you could add `fold` and accumulate a list of warnings to be printed out later.

Reply:
I tried to implement the fold-based approach. I hope it's somewhat similar to what you had in mind... any feedback would be much appreciated. 

One nice side effect that this change has is that now warnings are automatically sorted by line (at least in the examples I went over, not sure if it'll always hold), which I didn't expected.